src/mqtt/protocol/mqtt/mqtt_client.c

```diff
@@ -0,0 +1,1110 @@
+		packet_id = nni_mqtt_msg_get_pubrel_packet_id(msg);
+		work      = nni_id_get(&p->recv_unack, packet_id);
+		if (NULL == work) {
+			// ignore this message
+			nni_msg_free(msg);
+			nni_mtx_unlock(&s->mtx);
+			return;
+		}
+		// the transport handled sending the PUBCOMP for us
+		work_set_final(work);
+		nni_id_remove(&p->recv_unack, work->packet_id);
+		// ownership of work->msg to the lmq
+		mqtt_pipe_recv_msgq_putq(p, work->msg);
+		mqtt_run_recv_queue(s);
+		work->msg = msg;
+		mqtt_sock_free_work(s, work); // will release msg
+		nni_mtx_unlock(&s->mtx);
+		return;
+
+	case NNG_MQTT_PUBLISH:
+		// we have received a PUBLISH
+		qos = nni_mqtt_msg_get_publish_qos(msg);
+		if (2 > qos) {
+			// QoS 0, successful receipt
+			// QoS 1, the transport handled sending a PUBACK
+			mqtt_pipe_recv_msgq_putq(p, msg);
+			mqtt_run_recv_queue(s);
+			nni_mtx_unlock(&s->mtx);
+			return;
+		} else {
+			work = mqtt_sock_get_work(s);
+			if (work == NULL) {
+				nni_mtx_unlock(&s->mtx);
+				nni_pipe_close(p->pipe);
+				return;
+			}
+			work->qos = qos;
+			work->msg = msg; // keep the message
+			work->packet_id =
+			    nni_mqtt_msg_get_publish_packet_id(msg);
+			// the transport handled sending PUBREC,
+			// expect to receive a PUBREL
+			work_set_recv(work, NNG_MQTT_PUBREL);
+			nni_id_set(&p->recv_unack, work->packet_id, work);
+		}
+		break;
+
+	default:
+		// unexpected packet type, server misbehaviour
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
+		return;
+	}
+
+	if (work_is_error(work)) {
+		// protocol error, just close the connection
+		nni_mtx_unlock(&s->mtx);
+		nni_aio_finish_error(work->user_aio, NNG_EPROTO);
+		nni_pipe_close(p->pipe);
+		return;
+	} else if (work_is_final(work)) {
+		// good news, protocol state machine run to the end
+		nni_aio *aio = work->user_aio;
+		mqtt_sock_free_work(s, work);
+		nni_mtx_unlock(&s->mtx);
+		nni_aio_finish(aio, 0, 0);
+		return;
+	}
+
+	nni_mtx_unlock(&s->mtx);
+
+	return;
+}
+
+// Note: This routine should be called with the sock lock held.
+static void
+mqtt_recv_start(mqtt_sock_t *s, nni_aio *aio)
+{
+	mqtt_pipe_t *p = s->mqtt_pipe;
+	work_t *     work;
+
+	work = mqtt_sock_get_work(s);
+
+	if (NULL == work) {
+		nni_mtx_unlock(&s->mtx);
+		nni_aio_finish_error(aio, NNG_ENOMEM);
+		return;
+	}
+
+	work->user_aio = aio;
+	work_set_recv(work, NNG_MQTT_PUBLISH);
+	nni_list_append(&s->recv_queue, work); // enqueue to recv
+
+	// no open pipe
+	if (NULL == p || nni_atomic_get_bool(&p->closed)) {
+		return;
+	}
+
+	mqtt_run_recv_queue(s);
+}
+
+// Note: This routine should be called with the sock lock held.
+static void
+mqtt_run_recv_queue(mqtt_sock_t *s)
+{
+	work_t *     work = nni_list_first(&s->recv_queue);
+	mqtt_pipe_t *p    = s->mqtt_pipe;
+	nni_msg *    msg;
+
+	while (NULL != work) {
+		if (0 != nni_lmq_get(&p->recv_messages, &msg)) {
+			break;
+		}
+		nni_list_remove(&s->recv_queue, work);
+		// nni_pipe_recv(p->pipe, &work->recv_aio);
```

Comment:
This looks fishy.   Do you need to restart recv, or not?

Reply:
done